buildMix: fix: initial try (#112477)
NixOS/nixpkgs: doc/languages-frameworks/beam.section.md

## Introduction {#beam-introduction}

In this document and related Nix expressions, we use the term, _BEAM_, to describe the environment. BEAM is the name of the Erlang Virtual Machine and, as far as we're concerned, from a packaging perspective, all languages that run on the BEAM are interchangeable. That which varies, like the build system, is transparent to users of any given BEAM package, so we make no distinction.

## Structure {#beam-structure}

All BEAM-related expressions are available via the top-level `beam` attribute, which includes:

> - `interpreters`: a set of compilers running on the BEAM, including multiple Erlang/OTP versions (`beam.interpreters.erlangR19`, etc), Elixir (`beam.interpreters.elixir`) and LFE (Lisp-Flavored-Erlang) (`beam.interpreters.lfe`).

Reviewer: ```suggestion
- `interpreters`: a set of compilers running on the BEAM, including multiple Erlang/OTP versions (`beam.interpreters.erlangR19`, etc), Elixir (`beam.interpreters.elixir`) and LFE (Lisp Flavoured Erlang) (`beam.interpreters.lfe`).
```

#### Mix Packages {#mix-packages}

`mixRelease` is used to make a release in the mix sense. Dependencies will need to be fetched with `fetchMixDeps` and passed to it.

#### mixRelease - Elixir Phoenix example

Here is how your `default.nix` file would look.

```nix

> with import <nixpkgs> { };

Reviewer 2: That wouldn't work on nixos where nixpkgs can be called nixos.

Author: I was trying to take the most common case here.
What do you propose?
Should I add a comment like
`Replace nixpkgs with nixos depending on how your main channel is called`

Reviewer 2: > Replace nixpkgs with nixos if you are on NixOS. If your default channel is called something different you should insert its name instead.

Something like the above I think should be okay.

Author: Just added it

Author: Just added it

Reviewer 3: Wait, what? No. `nixpkgs` Nix path entry is always called `nixpkgs`, regardless of the channel name.

Author: Thanks for the comment. Today I learned something. Let me remove this.

Author: Ok should be good. Feel free to ping me, if you think it needs something else.

let
  packages = beam.packagesWith beam.interpreters.erlang;
  src = builtins.fetchgit {
    url = "ssh://git@github.com/your_id/your_repo";

>     rev = "replace_with_your_commit";

Reviewer 2: That should probably be fetchFromGitHub or at least use the anonymous https url.

Author: I used fetchgit because I suspect the majority of cases would be a private repo. My understanding is that the only way to fetch a private repo is with fetchgit and the ssh url.
Let me know if I'm wrong here.

Reviewer 2: To be honest I am not sure but this is fine for now.

  };

  pname = "your_project";
  version = "0.0.1";
  mixEnv = "prod";

  mixDeps = packages.fetchMixDeps {
    pname = "mix-deps-${pname}";
    inherit src mixEnv version;
    # nix will complain and tell you the right value to replace this with
    sha256 = lib.fakeSha256;
    # if you have build time environment variables add them here
    MY_ENV_VAR="my_value";
  };

>   nodeDependencies =
>     (pkgs.callPackage ./assets/default.nix { }).shell.nodeDependencies;

Reviewer 2: ```suggestion
  nodeDependencies = (pkgs.callPackage ./assets/default.nix { }).shell.nodeDependencies;
```

  frontEndFiles = stdenvNoCC.mkDerivation {

>     pname = "frontend-${name}";

Reviewer 2: ```suggestion
    pname = "frontend-${pname}";
```

otherwise that would be rebuild when the version updates but the deps are the same.

    nativeBuildInputs = [ nodejs ];

    inherit version src;

    buildPhase = ''
      cp -r ./assets $TEMPDIR

      mkdir -p $TEMPDIR/assets/node_modules/.cache
      cp -r ${nodeDependencies}/lib/node_modules $TEMPDIR/assets
      export PATH="${nodeDependencies}/bin:$PATH"

      cd $TEMPDIR/assets
      webpack --config ./webpack.config.js
      cd ..
    '';

    installPhase = ''
      cp -r ./priv/static $out/
    '';

    outputHashAlgo = "sha256";
    outputHashMode = "recursive";
    # nix will complain and tell you the right value to replace this with
    outputHash = lib.fakeSha256;

    impureEnvVars = lib.fetchers.proxyImpureEnvVars;
  };

in packages.mixRelease {
  inherit src pname version mixEnv mixDeps;
  # if you have build time environment variables add them here
  MY_ENV_VAR="my_value";
  preInstall = ''
    mkdir -p ./priv/static
    cp -r ${frontEndFiles} ./priv/static
  '';
}
```

> Setup will require the following steps

Reviewer 2: ```suggestion
Setup will require the following steps:
```

- Move your secrets to runtime environment variables. For more information refer to the [runtime.exs docs](https://hexdocs.pm/mix/Mix.Tasks.Release.html#module-runtime-configuration). On a fresh Phoenix build that would mean that both `DATABASE_URL` and `SECRET_KEY` need to be moved to `runtime.exs`.

> - `cd assets` and `node2nix --development` will generate a Nix expression containing your frontend dependencies

Reviewer 2: ```suggestion
- `cd assets` and `nix-shell -p node2nix --run node2nix --development` will generate a Nix expression containing your frontend dependencies
```

#### Mix Packages {#mix-packages}

`mixRelease` is used to make a release in the mix sense. Dependencies will need to be fetched with `fetchMixDeps` and passed to it.

#### releaseMix - Elixir Phoenix example

Here is how your `default.nix` file would look.

> ```default.nix

Reviewer 2: ```suggestion
```nix
```

with import <nixpkgs> { };

let
  packages = beam.packagesWith beam.interpreters.erlang;

>   src = builtins.fetchGit {

Reviewer: ```suggestion
  src = builtins.fetchgit {
```

  preInstall = ''
    mkdir -p ./priv/static
    cp -r ${frontEndFiles} ./priv/static
  '';
}
```

Setup will require the following steps

- Keep in mind that having secrets in the build time environment variables is not recommended. If those are not secrets they can be directly included in the config.
- Move your secrets to runtime environment variables. For more information about this check [runtime.exs docs](https://hexdocs.pm/mix/Mix.Tasks.Release.html#module-runtime-configuration). On a fresh Phoenix build that would mean that both of the `DATABASE_URL` and `SECRET_KEY` need be moved to `runtime.exs`.
- `cd assets` and `node2nix --development` will generate a Nix expression containing your frontend dependencies
- commit and push those changes
- you can now `nix-build .`
- When you need to run the release, you will need to at least supply the following environment variable
  - `RELEASE_TMP` is a directory that your program has write access to. it will be used to store the BEAM settings.

#### Example of creating a service for an Elixir - Phoenix project

In order to create a service with your release, you could add a `service.nix`
in your project with the following

> ```Nix

Reviewer 2: ```suggestion
```nix
```

in
{
  systemd.services.${release_name} = {
    wantedBy = [ "multi-user.target" ];
    after = [ "network.target" "postgresql.service" ];
    requires = [ "network-online.target" "postgresql.service" ];
    description = "my app";
    environment = {
      # RELEASE_TMP is used to write the state of the
      # VM configuration when the system is running
      # it needs to be a writable directory
      RELEASE_TMP = working_directory;
      # can be generated in an elixir console with
      # Base.encode32(:crypto.strong_rand_bytes(32))
      RELEASE_COOKIE = "my_cookie";
      MY_VAR = "my_var";
    };
    serviceConfig = {
      Type = "exec";
      DynamicUser = true;
      WorkingDirectory = working_directory;
      # Implied by DynamicUser, but just to emphasize due to RELEASE_TMP
      PrivateTmp = true;
      ExecStart = ''
        ${release}/bin/${release_name} start
      '';
      ExecStop = ''
        ${release}/bin/${release_name} stop
      '';
      ExecReload = ''
        ${release}/bin/${release_name} restart
      '';
      Restart = "on-failure";
      RestartSec = 5;
      StartLimitBurst = 3;
      StartLimitInterval = 10;
    };

>     # needed for disksup do have sh available

Reviewer: I'm not sure I understand this comment. Maybe the following accurate and clearer.
```suggestion
    # disksup requires bash
```

    path = [ pkgs.bash ];
  };

  environment.systemPackages = [ release ];
}
```

## How to Develop {#how-to-develop}

### Creating a Shell {#creating-a-shell}

> Usually, we need to create a `shell.nix` file and do our development inside of the environment specified therein. Just install your version of erlang and other interpreter, and then user your normal build tools. As an example with elixir:

Reviewer: ```suggestion
Usually, we need to create a `shell.nix` file and do our development inside of the environment specified therein. Just install your version of Erlang and any other interpreters, and then use your normal build tools. As an example with Elixir:
```

#### Example of creating a service for an Elixir - Phoenix project

In order to create a service with your release, you could add a `service.nix`
in your project with the following

```Nix
{config, pkgs, lib, ...}:

let
  release = pkgs.callPackage ./default.nix;
  release_name = "app";
  working_directory = "/home/app";
in
{
  systemd.services.${release_name} = {
    wantedBy = [ "multi-user.target" ];
    after = [ "network.target" "postgresql.service" ];
    requires = [ "network-online.target" "postgresql.service" ];
    description = "my app";
    serviceConfig = {
      Type = "exec";
      DynamicUser = true;
      WorkingDirectory = working_directory;
      # Implied by DynamicUser, but just to emphasize due to RELEASE_TMP
      PrivateTmp = true;
      ExecStart = ''
        ${release}/bin/${release_name} start
      '';
      ExecStop = ''
        ${release}/bin/${release_name} stop
      '';
      ExecReload = ''
        ${release}/bin/${release_name} restart
      '';
      Restart = "on-failure";
      RestartSec = 5;
      StartLimitBurst = 3;
      StartLimitInterval = 10;

>       environment = {

Reviewer 4: environment shouldn't be in serviceConfig

it needs to be directly in the service, like so: `systemd.services.${release_name}.environment = {}`

Author: Thanks!

        # RELEASE_TMP is used to write the state of the
        # VM configuration when the system is running
        # it needs to be a writable directory
        # if you don't set it, it will default to /tmp

>         RELEASE_TMP = working_directory;

Reviewer 5: Is this still needed with the new wrapper?

Author: The comment says that if you don't set it, it will default to /tmp. 
Maybe this is too wordy?
My personal opinion on this is that there is a reasonable default, but people would want to test this to something? At least having the possibility to override this clearly in the example made more sense to me.
Let me know if you feel differently, open to suggestions.

Reviewer 5: Oh yes, sorry I missed the comment... Never mind then!

Alternatively, we can use `buildHex` as a shortcut:

#### buildMix - Elixir Phoenix example

> Here is how your default.nix file would look like

Reviewer: ```suggestion
Here is how your `default.nix` file would look.
```

      cp -r ./assets $TEMPDIR

      mkdir -p $TEMPDIR/assets/node_modules/.cache
      cp -r ${nodeDependencies}/lib/node_modules $TEMPDIR/assets
      export PATH="${nodeDependencies}/bin:$PATH"

      cd $TEMPDIR/assets
      webpack --config ./webpack.config.js
      cd ..
    '';

    installPhase = ''
      cp -r ./priv/static $out/
    '';

    outputHashAlgo = "sha256";
    outputHashMode = "recursive";
    # nix will complain and tell you the right value to replace this with
    outputHash = lib.fakeSha256;

    impureEnvVars = lib.fetchers.proxyImpureEnvVars;
  };

in packages.buildMix {
  inherit src name version;
  mixEnv = "prod";
  # nix will complain and tell you the right value to replace this with
  depsSha256 = lib.fakeSha256;
  inherit src;
  # if you have build time environment variable add them here
  buildEnvVars = {
    MY_ENV_VAR="my_value";
  };
  preInstall = ''
    mkdir -p ./priv/static
    cp -r ${frontEndFiles} ./priv/static
  '';
}
```

Setup will require the following steps

- Keep in mind that having secrets in the build time environment variables is not recommended. If those are not secrets they can be directly included in the config.

> - move your secrets to runtime envionment variables. For more information about this check [runtime.exs docs](https://hexdocs.pm/mix/Mix.Tasks.Release.html#module-runtime-configuration). On a fresh Phoenix build that would mean that both of the `DATABASE_URL` and `SECRET_KEY` need be moved to `runtime.exs`.

Reviewer: ```suggestion
- Move your secrets to runtime environment variables. For more information about this check [runtime.exs docs](https://hexdocs.pm/mix/Mix.Tasks.Release.html#module-runtime-configuration). On a fresh Phoenix build that would mean that both of the `DATABASE_URL` and `SECRET_KEY` need be moved to `runtime.exs`.
```

> - `cd assets` and `node2nix --development` will generate a nix expression containing your frontend dependencies

Reviewer: ```suggestion
- `cd assets` and `node2nix --development` will generate a Nix expression containing your frontend dependencies
```

- commit and push those changes
- you can now `nix-build .`

> - When you need to run the release, you will need to at least supply the environment variable RELEASE_TMP that you have write access to.

Reviewer: ```suggestion
- When you need to run the release, you will need to at least supply the environment variable `RELEASE_TMP` that you have write access to.
```

Reviewer: The wording here feels awkward, but I don't have a suggestion at the moment.

- When you need to run the release, you will need to at least supply the following environment variables
  - `RELEASE_TMP` is a directory that your program has write access to. it will be used to store the BEAM settings.
  - `RELEASE_COOKIE` is the COOKIE that will be used by the beam to connect to other nodes.

#### Example of creating a service for an Elixir - Phoenix project

In order to create a service with your release, you could add a `service.nix`
in your project with the following

```Nix
{config, pkgs, lib, ...}:

let
  release = pkgs.callPackage ./default.nix;
  release_name = "app";
  working_directory = "/home/app";
in
{
  systemd.services.${release_name} = {
    wantedBy = [ "multi-user.target" ];
    after = [ "network.target" "postgresql.service" ];
    requires = [ "network-online.target" "postgresql.service" ];
    description = "my app";
    environment = {
      # RELEASE_TMP is used to write the state of the
      # VM configuration when the system is running
      # it needs to be a writable directory
      RELEASE_TMP = working_directory;
      # can be generated in an elixir console with
      # Base.encode32(:crypto.strong_rand_bytes(32))

>       RELEASE_COOKIE = "my_cookie";

Reviewer 5: I'm not sure those variables should be in the unit file: for the `RELEASE_TMP` variable, I think I recall it being useful for the `daemon` and `eval` commands, which can be used outside of the service. For the `RELEASE_COOKIE`, I don't think it should be world-readable in the nix store, and is also used outside of the service for controlling a running elixir project, for example. If you do not mind, since I will be making a PR for generically overriding the `runtime.exs` file, I can also add the wrapper that sets RELEASE_TMP and RELEASE_COOKIE in that same function, like I did partially [here](https://github.com/minijackson/nixpkgs/blob/9f34df8ff5997adc6301f897b070ab2ea357a64f/nixos/modules/services/web-apps/mobilizon.nix#L38-L39)

Alternatively, we can use `buildHex` as a shortcut:

#### buildMix - Elixir Phoenix example

Here is how your default.nix file would look like

```default.nix

> with import <nixpkgs-unstable> { };

Reviewer 6: ```suggestion
with import <nixpkgs> { };
```

once this hits stable channel, it shouldn't ask users to import unstable.

Author: Completely agree. Let's wait for it to hit stable channel. We wouldn't want the docs to be out of date in the meanwhile.

Reviewer 7: How would you remember to fix it? Also if you read the unstable manual than we can assume that nixpkgs is unstable. Also choosing `nixpkgs-unstable` as an channel name is completely arbitrary.

let
  packages = beam.packagesWith beam.interpreters.erlang;
  src = builtins.fetchGit {
    url = "ssh://git@github.com/your_id/your_repo";
    rev = "replace_with_your_commit";
  };

  nodeDependencies =
    (pkgs.callPackage ./assets/default.nix { }).shell.nodeDependencies;

in packages.buildMix {
  inherit src;
  name = "your_project";
  version = "0.0.1";
  mixEnv = "prod";
  # leave this empty and nix will complain and tell you the right value
  # to replace this with
  depsSha256 = "";
  inherit src;
  depsPreConfigure = ''
    # if you have build time environment variable add them here
    export MY_ENV_VAR="my_value";
  '';
  preConfigure = ''
    # if you have build time environment variable add them here

>     export MY_ENV_VAR="my_value";

Reviewer 6: Is this the same as the one in the section above? So build environment variables must be exported in both steps? You can make them available to all steps by sending them to the underlying mkDerivation as Nix variables such as `MY_ENV_VAR = "my_value";`. You could add an attribute set that you include in the args sent to mkDerivation to make it clearer.

Author: You're right. Adding an attribute called `build_env_vars = { MY_ENV_VAR = "my_value";};` then using a merge in the mkDerivation in both the build and the fetch steps to make them available.
Thanks!

Author: I added a buildEnvVars attribute. Trying to merge this and the default attrs resulted in an error, so I have removed the merging of the attrs with the rest of the variables for mkDerivation. What changes is that you can't override as easily steps for the derivartion. Here in particular, the preConfigure hook is not overridable anymore from passed attrs. It looses a bit of flexibility. Not sure how to proceed otherwise.

}
```

#### Elixir - Phoenix project

> Here is an example shell.nix

Author: Thanks!

}
```

#### Elixir - Phoenix project

> Here is an example shell.nix

Reviewer: ```suggestion
Here is an example `shell.nix`.
```

```shell.nix
# latest elixir version are only available on unstable

> with import <nixpkgs-unstable> { };

Reviewer 7: Same concern as above.

Author: I guess it makes sense to put it to just nixpkgs since by the time the doc is released, this should be available on stable.

}
```

#### Elixir - Phoenix project

Here is an example `shell.nix`.

```nix
with import <nixpkgs> { };

let
  # define packages to install
  basePackages = [
    git
    # replace with beam.packages.erlang.elixir_1_11 if you need
    beam.packages.erlang.elixir
    nodejs-15_x
    postgresql_13
    # only used for frontend dependencies
    # you are free to use yarn2nix as well
    nodePackages.node2nix
    # formatting js file
    nodePackages.prettier
  ];

>   inputs = basePackages ++ lib.optional stdenv.isLinux inotify-tools

Reviewer 2: ```suggestion
  inputs = basePackages ++ lib.optionals stdenv.isLinux [ inotify-tools ]
```

    ++ lib.optionals stdenv.isDarwin

>     (with darwin.apple_sdk.frameworks; [ CoreFoundation CoreServices ]);

Reviewer 2: That line wouldn't be valid in a nixpkgs package and ``CoreFoundation CoreServices`` would need to be inputs.

Author: I agree with you in the case of a nixpkgs package. However this file is just meant to be shell.nix example.
Do you still feel that we should recommend something different?

  # define shell startup command
  hooks = ''
    # this allows mix to work on the local directory

>     mkdir -p $PWD/.nix-mix
>     mkdir -p $PWD/.nix-hex

Reviewer 2: ```suggestion
    mkdir -p .nix-mix .nix-hex
```

    export MIX_HOME=$PWD/.nix-mix
    export HEX_HOME=$PWD/.nix-mix

>     export PATH=$MIX_HOME/bin:$PATH
>     export PATH=$HEX_HOME/bin:$PATH

Reviewer 2: ```suggestion
    export PATH=$MIX_HOME/bin:$HEX_HOME/bin:$PATH
```

Author: If it's okay with you, I would like to keep those 2 steps separated, I think it makes it easier to read.
Each step is only trying to do one thing.

>     # TODO: not sure how to make hex available without installing it
>     # afterwards.

Reviewer 2: ```suggestion
    # TODO: not sure how to make hex available without installing it afterwards.
```

    mix local.hex --if-missing
    export LANG=en_US.UTF-8
    export ERL_AFLAGS="-kernel shell_history enabled"

    # postges related
    # keep all your db data in a folder inside the project
    export PGDATA="$PWD/db"

    # phoenix related env vars
    export POOL_SIZE=15
    export DB_URL="postgresql://postgres:postgres@localhost:5432/db"
    export PORT=4000
    export MIX_ENV=dev
    # add your project env vars here

>     export API_KEY="your_api_key"

Reviewer 2: That adds it to the nix store which would make it world readable.

Author: Yes, I've just added it as an example. The idea was just to add an env_var example.
I'll replace with ENV_VAR="your_env_var"

}
```

#### Elixir - Phoenix project

Here is an example `shell.nix`.

> ```shell.nix

Reviewer 2: ```suggestion
```nix
```

    nodePackages.node2nix
    # formatting js file
    nodePackages.prettier
  ];

  inputs = basePackages ++ lib.optional stdenv.isLinux inotify-tools
    ++ lib.optionals stdenv.isDarwin
    (with darwin.apple_sdk.frameworks; [ CoreFoundation CoreServices ]);

  # define shell startup command
  hooks = ''
    # this allows mix to work on the local directory
    mkdir -p $PWD/.nix-mix
    mkdir -p $PWD/.nix-hex
    export MIX_HOME=$PWD/.nix-mix
    export HEX_HOME=$PWD/.nix-mix
    export PATH=$MIX_HOME/bin:$PATH
    export PATH=$HEX_HOME/bin:$PATH
    # TODO: not sure how to make hex available without installing it
    # afterwards.
    mix local.hex --if-missing
    export LANG=en_US.UTF-8
    export ERL_AFLAGS="-kernel shell_history enabled"

    # postges related
    # keep all your db data in a folder inside the project
    export PGDATA="$PWD/db"

    # phoenix related env vars
    export POOL_SIZE=15
    export DB_URL="postgresql://postgres:postgres@localhost:5432/db"
    export PORT=4000
    export MIX_ENV=dev
    # add your project env vars here
    export API_KEY="your_api_key"
  '';

in mkShell {
  buildInputs = inputs;
  shellHook = hooks;
}
```

> initializing the project will require the following steps

Reviewer: ```suggestion
Initializing the project will require the following steps:
```

- create the db directory `initdb ./db` (inside your mix project folder)
- create the postgres user `createuser postgres -ds`
- create the db `createdb db`
- start the postgres instance `pg_ctl -l "$PGDATA/server.log" start`

> - add the `/db` folder to your .gitignore

Reviewer: ```suggestion
- add the `/db` folder to your `.gitignore`
```
